### ISSUE TITLE
v0.5: RSA sigGen test responses not accepted by ACVP server due to "e"

### ISSUE THREAD

**Issue by smuellerDD** [usnistgov/ACVP#495]
When uploading of an RSA sigGen test result, the ACVP server says

```
{
  "acvVersion" : "0.5",
  "error" : "Incorrectly formatted JSON (659:4): Unknown or disallowed submission property: e"
}
```

My response looks like:

```
[
  {
    "acvVersion":"0.5"
  },
  {
    "vsId":176,
    "testGroups":[
      {
        "tgId":1,
        "tests":[
          {
            "tcId":1,
            "e":"a1324b6f",
            "n":"b77c206ecc545ab7c41a84048bae2f972aef78ade634d56a5fca2dc3a52c2ab13f7687c2fe4e7e8a0cc83e9540d531cc354725d32957def22c1a1a80ec8e76cc77c7ed4ea60fae2dc618dbd1e3f1b2a1bb216f4fd06175506fd05bf97b54a121468292727a86667035ad6ef2bc768616f050b9302308be8360061311ba6938b182de0bdf4c7dc91caf20211bc747d588b79ef3ab1a51903d90b8483262f2304b65a9da083dbd83eafb718ce5d77d580c8581e6b9ba143b14d12f6edd26c2cecaaf445fdfea2e48a23dd7339e0849d1bb767bd8691b93d19c464f0d6d88607778136b56c04886e3a95d367fe78af5da8314cfcbe1c9df0cb06ccc50aba9a4cb07",
            "signature":"272de94e4fc26dda47e2bedfc84209f822638a663ecced9d5af31e242e8729ef95820fee7a6e9aa8c601e2edee494f24ce16384cc6732727f559a8d9cea074d5bb8b5cf9f7ae3d68545b6cce2632b8962973859799a635c9d3bf22a95e6593482490b8282b9f21383beb5e916957c27d57a2fd6d232982c7cd0f80597f4f05a5d2cbf8c09793cb39481bd3879074512051ae5fa2117ee6efbceee9f6bb43f16bbb748ec1b2abefd5e48e72e06d8006443e73f35cbdb353c29173a5681c8bcedbace11d0449b6ff3128f902d0ef4992048515581e8e94a7bf8e96b4a79d9766fb1be2abd3fc0fb9e9a36c93c01302550a1bf28cb64f5717f14e9c14837299e536"
          },
```

What is wrong with the reporting of "e"?

**Comment by Kritner** [comment]
@smuellerDD it looks like RSA.SigGen is expecting `e`, and `n` values as test group level properties that should be shared for the entirety of the group.  I guess the spec needs updating related to this?  I guess there's a few other RSA spec related issues out there as well.

**Comment by smuellerDD** [comment]
How is that supposed to work? Test groups is an array.

**Comment by celic** [comment]
It would look like this...
```
[
  {
    "acvVersion":"0.5"
  },
  {
    "vsId":176,
    "testGroups":[
      {
        "tgId":1,
        "e":"a1324b6f",
        "n":"b77c206ecc545ab...",
        "tests":[
          {
            "tcId":1,
            "signature":"272de94e4fc26..."
          },
```

**Comment by celic** [comment]
The idea is to allow IUTs to re-use the key for generating a bunch of signatures. We already test the ability to generate keys under KeyGen.

**Comment by smuellerDD** [comment]
Am Donnerstag, 8. November 2018, 16:45:05 CET schrieb Russ Hammett:

Hi Russ,

> @smuellerDD it looks like RSA.SigGen is expecting `e`, and `n` values as
> test group level properties that should be shared for the entirety of the
> group.  I guess the spec needs updating related to this?  I guess there's a
> few other RSA spec related issues out there as well.

The spec needs updating:

- e / n are in table 12 - for siggen they need to somehow be separated out

- examples need to be updated.

But in any case, may I ask about why the change in JSON structure? I could 
happily live with the n/e values being listed along with the signature. The 
test harness can (and I already did) generate only one key per modulo and 
reuse it. I.e. previously the e/n values were therefore duplicated for all 
signatures with the same modulo.

IMHO that would keep thinks easier, because the response structure for all RSA 
answers where you need n/e would be identical.

Ciao
Stephan

**Comment by celic** [comment]
It avoids unnecessary duplicate data. As well it allows the server to better understand what is happening inside of the test group. The server (as of now) doesn't have a way of looking at multiple test cases simultaneously. If the key were kept in each test case, IUTs could use different keys per case, or the same key without the server being able to easily tell. While this isn't necessarily a bad thing, preventing this helps keep a consistent approach to the test cases and encourages faster processing. 

The spec does need updating. I think it was updated to reflect this change, but honestly the tables are incredibly hard to understand when each value is applicable. See the symmetric spec for how we plan on changing this... We just haven't gotten to RSA yet.

**Comment by elliemdaw** [comment]
I am working on adding this to Cisco's client... however wanted to revisit a little bit. Currently, if it is the responsibility of the crypto module to generate values, these values are unique to each test case. However with these changes, this requires the application to keep state between different test cases... I understand the performance rationale, but this behavior seems a little odd?

**Comment by smuellerDD** [comment]
I agree that the test response is a departure from all other response structures.

But other JSON definitions from the ACVP have the structure defined here. So, I am not sure that the difference to other JSON responses can serve as an argument.

**Comment by awolf-bc** [comment]
I just ran into this issue after reading the current RSA SPEC file. Luckily, I had worked on the DSA 'sigGen' code before. That algorithm SPEC places the 'e' and 'n' values in the 'group' object of the RSP JSON.

Can the RSA SPEC be updated, soon? Other people might have a harder time figuring this out...
